vendor/resenje.org/singleflight/singleflight.go

```diff
@@ -60,30 +85,45 @@ func (g *Group[K, V]) Do(ctx context.Context, key K, fn func(ctx context.Context
 	g.mu.Unlock()
 
 	go func() {
+		defer func() {
+			if v := recover(); v != nil {
+				c.panicErr = &panicError{value: v, stack: debug.Stack()}
+			}
+			close(c.done)
+		}()
+
 		c.val, c.err = fn(callCtx)
-		close(c.done)
 	}()
 
 	return g.wait(ctx, key, c)
 }
 
 // wait for function passed to Do to finish or context to be done.
 func (g *Group[K, V]) wait(ctx context.Context, key K, c *call[V]) (v V, shared bool, err error) {
+	var panicErr *panicError
 	select {
 	case <-c.done:
 		v = c.val
 		err = c.err
+		panicErr = c.panicErr
 	case <-ctx.Done():
 		err = ctx.Err()
 	}
 	g.mu.Lock()
 	c.counter--
 	if c.counter == 0 {
 		c.cancel()
-		delete(g.calls, key)
+		if !c.forgotten {
+			delete(g.calls, key)
+		}
 	}
 	shared = c.shared
 	g.mu.Unlock()
+
+	if panicErr != nil {
+		panic(panicErr)
+	}
```

Comment:
Is there any situation where we could trip up a panic now, or would that already be such an exception that it already panicked through other ways, or a panic is the only right thing to do?

Reply:
The singleflight behavior for the same key _should_ produce the same result, so it's very likely that the panic would happen again.

The `x/sync/singleflight` also does the same thing: https://cs.opensource.google/go/x/sync/+/refs/tags/v0.9.0:singleflight/singleflight.go;l=102